internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,207 @@
+// size.
+type logGatherer struct {
+	targetName string
+	tickPeriod time.Duration
+	cancel     chan struct{}
+	client     logClient
+
+	entryCh     chan servicelog.Entry
+	pullers     map[string]*logPuller
+	pullersLock sync.Mutex
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	return &logGatherer{
+		targetName: target.Name,
+		tickPeriod: 1 * time.Second,
+		cancel:     make(chan struct{}),
+		client:     client,
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    map[string]*logPuller{},
+	}, nil
+}
+
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	newPullers := map[string]*logPuller{}
+
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		if puller, ok := g.pullers[service.Name]; ok {
+			// Move over existing puller
+			newPullers[service.Name] = puller
+			delete(g.pullers, service.Name)
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for svc, puller := range g.pullers {
+		err := puller.tomb.Killf("gatherer closed")
+		if err != nil {
+			logger.Noticef("Error shutting down puller for target %q, service %q: %v",
+				g.targetName, svc, err)
+		}
+	}
+	g.pullers = newPullers
+}
+
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+	g.pullers[service.Name] = newLogPuller(g.entryCh, buffer)
+}
+
+func (g *logGatherer) loop() {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// Timeout - flush
+			err := g.client.Flush()
+			if err != nil {
+				logger.Noticef("Cannot send logs to target %q: %v", g.targetName, err)
+			}
+
+		case <-g.cancel:
+			// Gatherer has been stopped - flush any remaining logs
+			err := g.client.Flush()
+			if err != nil {
+				logger.Noticef("Cannot send logs to target %q: %v", g.targetName, err)
+			}
+			return
+
+		case entry := <-g.entryCh:
+			err := g.client.Write(entry)
+			if err != nil {
+				logger.Noticef("Cannot write logs to target %q: %v", g.targetName, err)
+			}
+		}
+	}
+}
+
+// stop closes the cancel channel, thereby terminating the main loop.
+func (g *logGatherer) stop() {
+	close(g.cancel)
+
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("gatherer closed")
+	}
+}
```

Comment:
we need to force the client to flush here, with a timeout.

Reply:
To have a timeout for the client flush here, we'd need to pass a channel into the `logClient.Flush` method and ask client implementations to respect that